doc/CONFIG.md
# Configuration

The configuration use a format somewhere betweem TOML and ini, they use the
`.ini` file extension for syntax highlighting.

## Format

The basic format is just like ini:

```ini
[section]
key = value

# Sections can be nested
# either like this:
[nested.section]
key = value

# or like this:
[nested]
# ...
[.section]  # leading '.' appends the name to the previous section
key = value
```

The difference is that we have set value types for the various things we need.

## Types

identifies | description | examples
---|---|---
`bool` | `true` or `false` |
`uint` | a unsigned decimal integer | `21`
`float` | a real number | `3.141`
`size` | like `float` but may be be suffixed by one of `px`, `mm`, `cm`, `em`, `%`. | See following rows
&nbsp; | `px` or no prefix: number of pixels | `10`
&nbsp; | `mm` or `cm`: physical size in millimeters/centimeters, based on monitor DPI | `2mm`, `1cm`
&nbsp; | `em`: multiple of font size, only available for certain values* | `1.1em`
&nbsp; | `%`: relative to some superior element, only available for certain values* | `80%`
`string` | a delimited string | `'single quoted'`, `"double quoted"`
`font` | alias for `string` | `'sans 14'`
`alignment` | `left`, `center`, or `right` |
`[T]` | a list where `T` is one of the other types. This is the only value that can span multiple lines. | `['minimize', 'maximize', 'close']`, `(16, 9)`
`[T; N]` | a list of `T` that must have `N` elements | `[1mm, 1mm, 1mm, 1mm]`
`color` | a color value or link | See following rows
&nbsp; | `#RRGGBB` or `#RRGGBBAA`: a hex color value | `#428BCA`, `#161514CC`
&nbsp; | `rgb(r, g, b)` or `rgba(r, g, b, a)`: components given as `uint`s | `rgb(235, 64, 52)`, `rgba(22, 21, 20, 204)`
&nbsp; | `foo.bar`: link to another element | `window.buttons.close`

*: Size values will indicate when they can be relative to font/superior element sizes.

## Sections

### [general]

Key and type | Description | Default
---|---|---
`meta_window_classes` <br> `[string]` | classes to use as meta windows; these windows are visible from each workspace, are undecorated, and cannot be interacted with. Additionally, windows with the title window_manager_bar are also meta windows. | `[]`
`default_notification_timeout` <br> `uint` | default time (milliseconds)desktop notifications are displayed for if no time is specified, if set to 0 notifications never expire. | `6000`
`double_click_time` <br> `uint` | maximum time after which a second click is considered a double-click in milliseconds. | `500`
`grid_resize` <br> `bool` | run [grid-resize](https://github.com/JaMo42/grid-resize) when holding the left and right mouse buttons while moving a window. | `false`
`grid_resize_grid_size` <br> `[uint; 2]` | the vertical and horizontal columns for grid-resize. | `(16, 9)`
`grid_resize_live` <br> `bool` | run grid-size in live mode. | `false`
`scale_base_fonts` <br> `bool` | see [fonts](#fonts). | `true`

### [layout]

Key and type | Description | Default
---|---|---
`workspaces` <br> `uint` | number of workspaces. | `1`
`gaps` <br> `size` | width of inner gaps for snapped windows. | `0`
`pad` <br> `[uint; 4]` | padding from screen edges for snapped windows, values are [Top, Bottom, Left, Right]. This sets the padding for the main monitor. If the bar is enabled its height is added to the top padding. | `[0, 0, 0, 0]`
`secondary_pad` <br> `[uint; 4]` | like `pad` but for all non-primary monitors. | `[0, 0, 0, 0]`
`smart_window_splacement` <br> `bool` | Try to place new clients so they don't overlap any existing clients or at least have as little overlap as possible. | `true`
`smart_window_placement_max` <br> `unint` | Do not attemp to do smart window placement if there are already this many clients on the main monitor. `0` means no limit. | `0`

### [window]

Key and type | Description | Default
---|---|---
`border` <br> `size` | size of border around windows | `1mm`
`title_font` <br> `font` | &nbsp; | `'sans'`
`title_font_size` <br> `uint` | see [Fonts](#fonts) | `14`
`title_font_scaling_percent` <br> `uint` | &nbsp; | `100`
`title_bar_height` <br> `size`, relative to title font size | height of the title bar | `1.1em`
`title_alignment` <br> `alignment` | window title alignment | `left`
`left_buttons` <br> `[string]` | left window buttons, valid values are `minimize`, `maximize`, and `close` | `[]`
`right_buttons` <br> `[string]` | &nbsp; | `['close']`
`icon_size` <br> `size`, relative to title bar height | size of window icons. If `0` window icons are disabled. | `80%`
`button_icon_size` <br> `size`, relative to title bar height | size of button icons | `75%`
`circle_buttons` <br> `bool` | enable [circular window buttons](#circular-buttons) | `false`
`extend_frame` <br>  `sizeo` | how far to extend the clickable window frame for resizing. `0` means disable frame extension. | `1mm`

### [theme]

Key and type | Description | Default
---|---|---
`colors` <br> `string` | name of the [color scheme](#color-scheme) | `'default'`
`icons` <br> `string` | name of the icon theme | `'Papirus'`

### [keys]

Key and type | Description | Default
---|---|---
`super` <br> `string` | the user-defined modifier. This is used for mouse key bindings as well as the `Mod` modifier in key bindings. | `'Super'`

### [keys.bindings]

See [Key bindings](#key-bindings).

### [bar]

Key and type | Description | Default
---|---|---
`height` <br> `size`, relative to font size | &nbsp; | `1.1em`
`font` <br> `font` | &nbsp; | `'sans 14'`
`time_format` <br> `string` | format for the date-time widget, uses strftime format. | `'%a %b %e %H:%M %Y'`
`localized_time` <br> `bool` | use the `LC_TIME` locale for the format values in `time_format` | `true`
`power_supply` <br> `string` | the power supply to use for the battery widget | `'BAT0'`
`update_interval` <br> `uint` | time between automatic bar updates in milliseconds | `10000`
`volume_mixer_title_width` <br> `size`, relative to font size, main monitor width, and main monitor dpmm | width for the program name column in the volume mixer | `15%`
`volume_mixer_grouping` <br> `"name"` or `"pid"` | specify if applications in the volume mixer are grouped by name or process id, invalid values result in `"pid"` being used. Note that the process id is that provided by pulse audio and for example browsers tend to have the same process id for all windows. If different instances already have different volume levels they are not grouped. | `"name"`
`window_switcher_sizes` <br> `[uint]` | Percentages of the main monitor height to try for the window switcher window preview height (the height of the actual window preview, excluding the title). | `[30, 20, 15, 10]`

### [dock]

Key and type | Description | Default
---|---|---
`height` <br> `size`, relative to main monitor height | &nbsp; | `10%`
`pinned` <br> `[string]` | pinned dock items, these are always in the dock even if they have no client | `[]`
`focused_client_on_top` <br> `bool` | move the focused instance of an application to the top of the list in context menus | `false`
`focused_urgent` <br> `bool` | if there is a urgent client, focus it instead of the active client when clicking a dock item | `false`
`item_size` <br> `size`, relative to dock height | size of dock items | `80%`
`icon_size` <dr> `size`, relative to item size | size of icons on items | `85%`
`context_show_workspaces` <br> `bool` | show the workspace of clients in context menus, unless all clients are on the current workspace | `true`
`auto_indicator_colors` | for programs with icons, use the average color of the icon as the indicator color. | `true`

### [split_handles]

Key and type | Description | Default
---|---|---
`size` <br> `size` | width of split handles | `2mm`
`vertical_sticky` <br> `[uint]` | percentages of the monitor width where split handles will stick to while moving | `[50]`
`horizontal_sticky` <br> `[uint]` | percentages of the monitor height where split handles will stick to while moving | `[50]`
`min_split_size` <br> `uint` | percentage of minimum split size based on monitor size | `10`

## Fonts

- `window.title_font_size`: this is the size the window title font uses on the primaty monitor, for all other monitors this is scaled to have the same visual size. Set to `0` to disable this feature and use the size specified in the `window.title_font` property.

- `window.tile_font_scaling_percent`: percentage that he title font scaling will take effect, since on some resolution and DPI combinations fonts may lose clarity at full scaling.

- `general.scale_base_fonts`: when set to `true`, the base size of all fonts is scaled to have the same visual size when the DPI of the primary monitor changes. This value is also affected by the `window.tile_font_scaling_percent` value.

## Key bindings

Example:

```ini
[keys]
[.bindings]
Mod+Shift+Q = quit
Mod+space = $ launcher
... = $ command 'one argument'
... = $ command "one 'argument'"
... = $ command "one \\"argument\\""
... = $ command one\ argument
```

In this section the keys are a `+`-separated list of 0 or more modifiers and a key.

The key names must adhere to the requirements of `XStringToKeysym`, that is:
```
Standard KeySym names are obtained from <X11/keysymdef.h> by removing the XK_ prefix from each name.
KeySyms that are not part of the Xlib standard also may be obtained with this function.
The set of KeySyms that are available in this manner and the mechanisms by which Xlib obtains them is implementation-dependent.
```

If a value starts with a `$` it's a command that should be run if the key is pressed.

Otherwise it must be one of the [actions](#actions).

## Actions

TODO, here are the names:

- `quit`
- `quit_dialog`
- `snap_left`
- `snap_right`
- `snap_up`
- `snap_down`
- `maximize`
- `unsnap_or_center`
- `close_window`
- `raise_all`
- `decrease_volume`
- `increase_volume`
- `mute_volume`
- `move_to_next_monitor`
- `move_to_prev_monitor`

## Color scheme

Color schemes are located at `colors/<name>.ini` relative to the config file.

Example:

```ini
[palette]
named_color = rgb(12, 34, 56)

[window]
focused = #456789
urgent = windows.buttons.close
#        Links to non-palette colors use full dotted paths

[bar]
background = named_color
#            Links to palette colors just use their names
```

All values are of type `color`.

### [palette]

This section is not used by anything and just contains colors to use in other sections.

If can contain any key.

### [misc]

Element | Description | Default
---|---|---
`background` | color of the root window if no other program sets the wallpaper | `#000000`

### [window]

Element | Description | Default
---|---|---
`focused`
`focused_text`
`normal`
`normal_text`
`selected`
`selected_text`
`urgent`
`urgent_text`

### [window.buttons]

Element | Description | Default
---|---|---
`close`
`close_hovered`
`maximize`
`maximize_hovered`
`minimize`
`minimize_hovered`

### [bar]

Element | Description | Default
---|---|---
`background`
`text`
`workspace`
`workspace_text`
`active_workspace`
`active_workspace_text`
`urgent_workspace`
`urgent_workspace_text`

### [dock]

Element | Description | Default
---|---|---
`background`
`hovered`
`urgent`
`indicator`

### [tooltip]

Element | Description | Default
---|---|---
`background`
`text`

### [context_menu]

Element | Description | Default
---|---|---
`background`
`text`

### [notifications]

Element | Description | Default
---|---|---
`background`
`text`
`divider`

## Circular buttons

If circle buttons are enabled, `<button>` becomes the circle color for normal windows and `<button>_hovered` becomes the circle color for focused windows. The color of the actual icon gets derived from the `<button>_hovered` color.


src/config.rs
use crate::{
    action,
    cfg::{gen::parsed_config, parse::Parser},
    client::Client,
    color_scheme::ColorScheme,
    config_types::ColorSchemeParser,
    draw::{Alignment, DrawingContext},
    error::OrFatal,
    layout::{lerp, ClientLayout, LayoutClass},
    monitors::{monitors_mut, WindowAreaPadding},
    paths,
    window_manager::WindowManager,
    x::{string_to_keysym, Display, ModifierMapping},
    AnyResult,
};
use pango::FontDescription;
use std::{
    cell::RefCell,
    collections::{hash_map, HashMap},
    rc::Rc,
    sync::Arc,
};
use x11::keysym;
use xcb::x::{KeyButMask, Keycode, ModMask};

#[derive(Eq, PartialEq, Hash, Copy, Clone, Debug)]
pub struct Key {
    modifiers: ModMask,
    code: Keycode,
}

impl Key {
    pub fn from_str(display: &Display, s: &str, modifiers: ModMask) -> Option<Self> {
        string_to_keysym(s).map(|sym| Self::from_sym(display, sym as u32, modifiers))
    }

    pub fn from_sym(display: &Display, sym: u32, modifiers: ModMask) -> Self {
        Self {
            modifiers,
            code: display.keysym_to_keycode(sym),
        }
    }

    pub fn modifiers(&self) -> ModMask {
        self.modifiers
    }

    pub fn code(&self) -> u8 {
        self.code
    }
}

fn str2mod(s: &str, user_mod: ModMask, modmap: &ModifierMapping) -> ModMask {
    match s.to_ascii_lowercase().as_str() {
        "win" => modmap.win(),
        "shift" => modmap.shift(),
        "alt" => modmap.alt(),
        "ctrl" => modmap.control(),
        "mod" => user_mod,
        _ => ModMask::empty(),
    }
}

/// Parses a `+` separated sequence of modifiers. The user mod is defaulted to
/// to super/win key.
fn modifiers_from_string(s: &str, modmap: &ModifierMapping) -> ModMask {
    let mut mods = ModMask::empty();
    for mod_str in s.split('+') {
        mods |= str2mod(mod_str, modmap.win(), modmap);
    }
    mods
}

fn mods_and_key_from_string<'a>(
    s: &'a str,
    user_mod: ModMask,
    modmap: &ModifierMapping,
) -> (ModMask, &'a str) {
    let mut mods = ModMask::empty();
    let mut key = "";
    let mut it = s.split('+').peekable();
    while let Some(i) = it.next() {
        if it.peek().is_some() {
            mods |= str2mod(i, user_mod, modmap);
        } else {
            key = i;
        }
    }
    (mods, key)
}

#[derive(Copy, Clone)]
pub struct WorkspaceAction(
    /// function: fn(window_manager, workspace_index, maybe_client)
    pub fn(&WindowManager, usize, Option<&Client>),
    /// `workspace_index` argument for the function
    pub usize,
    /// whether this action requires a focused client
    pub bool,
);

#[derive(Clone)]
pub enum Action {
    Client(fn(&Client)),
    Workspace(WorkspaceAction),
    Launch(Vec<String>),
    Generic(fn(&WindowManager)),
}

#[derive(Default)]
pub struct KeyBindings {
    pairs: Vec<(String, Action)>,
}

impl KeyBindings {
    pub fn push(&mut self, mods_and_key: String, action: Action) {
        self.pairs.push((mods_and_key, action));
    }

    fn into_map(
        self,
        display: &Display,
        user_mod: ModMask,
        modmap: &ModifierMapping,
    ) -> HashMap<Key, Action> {
        let mut map = HashMap::new();
        for (mods_and_key, action) in self.pairs.into_iter() {
            let (mods, key) = mods_and_key_from_string(&mods_and_key, user_mod, modmap);
            // Validity of key was already checked during parsing so unwrap is
            // safe here.
            let key = Key::from_str(display, key, mods).unwrap();
            map.insert(key, action);
        }
        map
    }
}

#[derive(Copy, Clone, Debug)]
pub enum Size {
    /// Direct pixel count
    Pixels(u16),
    /// Millimeters.
    Physical(f64),
    /// Percentage of a superior element.
    Percent(f64),
    /// Size relative to some font size.
    PercentOfFont(f64),
}

impl Size {
    /// Creates a new pixel size value.
    pub fn new(pixels: u16) -> Self {
        Self::Pixels(pixels)
    }

    /// Constructs a new size value based on the given suffix.
    pub fn new_with_suffix(num: f64, suffix: &str) -> Self {
        match suffix {
            "" | "px" => Self::Pixels(num.round() as u16),
            "mm" => Self::Physical(num),
            "cm" => Self::Physical(num * 10.0),
            "%" => Self::Percent(num.clamp(0.0, 100.0) / 100.0),
            "em" => Self::PercentOfFont(num.max(1.0)),
            x => panic!("invalid suffix for size: {x}"),
        }
    }

    /// Tries to resolve the size to a pixel value.
    pub fn resolve(
        &self,
        dpmm: Option<f64>,
        relative_to: Option<u16>,
        font_height: Option<u16>,
    ) -> u16 {
        match *self {
            Self::Pixels(pixels) => pixels,
            Self::Physical(mm) => (dpmm.unwrap() * mm).round() as u16,
            Self::Percent(p) => (relative_to.unwrap() as f64 * p).round() as u16,
            Self::PercentOfFont(p) => (font_height.unwrap() as f64 * p).round() as u16,
        }
    }
}

fn parse_color_scheme(display: &Display, name: &str) -> AnyResult<ColorScheme> {
    log::trace!("Loading color scheme");
    let pathname = format!("{}/{}.ini", paths::colors_dir(), name);
    let mut parser = Parser::new(display, &pathname)?;
    let mut scheme_parser = ColorSchemeParser::default();
    parser.parse(&mut scheme_parser);
    scheme_parser.finish()
}

/// Tries to find the full path to the given icon theme name.
fn find_icon_theme(name: &str) -> AnyResult<String> {
    let home = std::env::var("HOME")?;
    let directories = [
        "/usr/share/icons".to_string(),
        format!("{}/{}", home, ".local/share/icons"),
        format!("{}/{}", home, ".icons"),
    ];
    log::trace!("Looking for icon theme");
    for d in directories {
        let path = format!("{}/{}", d, name);
        log::trace!(" - {}", path);
        if std::fs::metadata(&path).is_ok() {
            log::trace!(" -> found");
            return Ok(path);
        }
    }
    log::trace!(" -> none found");
    Err(format!("Icon theme not found: {}", name))?
}

parsed_config! {
    sections => {
        General {
            meta_window_classes: Vec<String> = "[]",
            default_notification_timeout: u64 = "6000",
            double_click_time: u32 = "500",
            grid_resize: bool = "false",
            grid_resize_grid_size: (u16, u16) = "[16, 9]",
            grid_resize_live: bool = "false",
            scale_base_fonts: bool = "true",
            window_switcher_sizes: Vec<u32> = "[30, 20, 15, 10]",
            window_switcher_filter: cairo::Filter = "Good",
        }
        Layout {
            workspaces: usize = "1",
            gaps: Size = "0",
            padding: WindowAreaPadding = "[0, 0, 0, 0]",
            secondary_padding: WindowAreaPadding = "[0, 0, 0, 0]",
            smart_window_placement: bool = "true",
            smart_window_placement_max: usize = "0",
        }
        Window {
            border: Size = "1mm",
            title_font: FontDescription = "'sans'",
            title_font_size: u16 = "14",
            title_font_scaling_percent: u16 = "100",
            title_bar_height: Size = "1.1em",
            title_alignment: Alignment = "left",
            left_buttons: Vec<String> = "[]",
            right_buttons: Vec<String> = "['close']",
            icon_size: Size = "80%",
            button_icon_size: Size = "75%",
            circle_buttons: bool = "false",
            extend_frame: Size = "1mm",
        }
        Theme {
            colors: String = "'default'",
            icons: String = "'Papirus'",
        }
        Keys {
            modifier: String = "'Super'",
        }
        Bar {
            enable: bool = "true",
            height: Size = "1.1em",
            font: FontDescription = "'sans 14'",
            time_format: String = "'%a %b %e %H:%M %Y'",
            localized_time: bool = "true",
            power_supply: String = "'BAT0'",
            update_interval: u64 = "10000",
            volume_mixer_title_width: Size = "15%",
            volume_mixer_grouping: String = "'name'",
        }
        Dock {
            enable: bool = "true",
            height: Size = "10%",
            offset: Size = "0",
            pinned: Vec<String> = "[]",
            focused_client_on_top: bool = "false",
            focus_urgent: bool = "false",
            item_size: Size = "80%",
            icon_size: Size = "85%",
            context_show_workspaces: bool = "true",
            auto_indicator_colors: bool = "true",
        }
        SplitHandles {
            size: Size = "2mm",
            vertical_sticky: Vec<u16> = "[50]",
            horizontal_sticky: Vec<u16> = "[50]",
            min_split_size: u16 = "10",
        }
    }
    config => {
        general: General => "general",
        layout: Layout => "layout",
        window: Window => "window",
        theme: Theme => "theme",
        keys: Keys => "keys",
        key_bindings: KeyBindings => "keys.bindings",
        bar: Bar => "bar",
        dock: Dock => "dock",
        split_handles: SplitHandles => "split_handles",
    }
}

pub struct Config {
    key_binds: HashMap<Key, Action>,
    modifier_str: String,
    modifier: RefCell<ModMask>,
    client_layout: Rc<RefCell<LayoutClass<ClientLayout>>>,
    pub general: General,
    pub layout: Layout,
    pub window: Window,
    pub bar: Bar,
    pub dock: Dock,
    pub split_handles: SplitHandles,
    pub colors: ColorScheme,
    pub icon_theme: String,
}

unsafe impl Sync for Config {}

impl Config {
    pub fn try_load(display: &Arc<Display>, dc: &DrawingContext) -> AnyResult<Self> {
        let mut parsed = ParsedConfig::default();
        log::trace!("Loading configuration");
        Parser::new(display, paths::config_path().as_str())?.parse(&mut parsed);
        let mut modmap = ModifierMapping::default();
        modmap.refresh(display);
        let user_mod = modifiers_from_string(&parsed.keys.modifier, &modmap);
        let mut this = Self {
            key_binds: parsed.key_bindings.into_map(display, user_mod, &modmap),
            modifier_str: parsed.keys.modifier,
            modifier: RefCell::new(user_mod),
            client_layout: Rc::new(RefCell::new(LayoutClass::default())),
            general: parsed.general,
            layout: parsed.layout,
            window: parsed.window,
            bar: parsed.bar,
            dock: parsed.dock,
            split_handles: parsed.split_handles,
            colors: parse_color_scheme(display, &parsed.theme.colors)?,
            icon_theme: find_icon_theme(&parsed.theme.icons)?,
        };
        for ws_idx in 0..this.layout.workspaces {
            let sym = keysym::XK_1 + ws_idx as u32;
            this.key_binds.insert(
                Key::from_sym(display, sym, user_mod),
                Action::Workspace(WorkspaceAction(action::select_workspace, ws_idx, false)),
            );
            this.key_binds.insert(
                Key::from_sym(display, sym, user_mod | modmap.shift()),
                Action::Workspace(WorkspaceAction(action::move_to_workspace, ws_idx, true)),
            );
        }
        this.key_binds.insert(
            Key::from_sym(display, keysym::XK_Tab, modmap.alt()),
            Action::Generic(action::switch_window),
        );
        this.recompute_layouts(dc);
        monitors_mut().set_window_areas(&this.layout.padding, &this.layout.secondary_padding);
        this.general
            .meta_window_classes
            .retain(|class| !class.is_empty());
        // todo: verify all relative size values are valid.
        Ok(this)
    }

    pub fn load(display: &Arc<Display>, dc: &DrawingContext) -> Self {
        Self::try_load(display, dc).unwrap_or_fatal(display)
    }

    pub fn modifier(&self) -> ModMask {
        *self.modifier.borrow()
    }

    pub fn refresh_modifier(&self, modmap: &ModifierMapping) {
        *self.modifier.borrow_mut() = str2mod(&self.modifier_str, modmap.win(), modmap);
    }

    pub fn get_key_binding(&self, code: u8, modifiers: KeyButMask) -> Option<Action> {
        let mod_bits = modifiers.bits() & ModMask::all().bits();
        let modifiers = unsafe { ModMask::from_bits_unchecked(mod_bits) };
        self.key_binds.get(&Key { modifiers, code }).cloned()
    }

    pub fn iter_keys(&self) -> hash_map::Keys<Key, Action> {
        self.key_binds.keys()
    }

    pub fn recompute_layouts(&self, dc: &DrawingContext) {
        self.client_layout.replace(LayoutClass::new(self, dc));
    }

    pub fn scale_fonts(&self, factor: f64) {
        fn scale_font(font: &mut FontDescription, factor: f64) {
            font.set_size((font.size() as f64 * factor).round() as i32)
        }
        let percent = self.window.title_font_scaling_percent as f64 / 100.0;
        let factor = lerp(1.0, factor, percent);
        log::trace!("Scaling fonts by {}%", (factor * 100.0).round() as u16);
        // These were originally not supposed to be mutable and I don't want
        // to put them all in cell types. This should be fine.
        #[allow(clippy::cast_ref_to_mut)]
        let this = unsafe { &mut *(self as *const Self as *mut Self) };
        scale_font(&mut this.bar.font, factor);
        if this.window.title_font_size > 0 {
            this.window.title_font_size =
                (this.window.title_font_size as f64 * factor).round() as u16;
        } else {
            scale_font(&mut this.window.title_font, factor);
        }
    }

    pub fn client_layout(&self) -> Rc<RefCell<LayoutClass<ClientLayout>>> {
        self.client_layout.clone()
    }
}
